File: internal/actions/assign.go
package actions

import (
	"context"
	"errors"
	"fmt"
	"sort"

	"github.com/google/uuid"
	"github.com/jackc/pgx/v5"

	"github.com/openforms/openforms/internal/auth"
	"github.com/openforms/openforms/internal/db"
	"github.com/openforms/openforms/internal/jobs"
	"github.com/openforms/openforms/internal/submissions"
)

func (r *runner) assign(ctx context.Context, job jobs.Job) error {
	ac, err := r.load(ctx, job)
	if err != nil {
		return err
	}
	a := ac.payload.Action
	var target auth.User
	switch {
	case a.User != "":
		u, err := r.d.Auth.GetUserByEmail(ctx, job.OrgID, a.User)
		if errors.Is(err, auth.ErrNotFound) {
			return jobs.Permanent(fmt.Errorf("assign: no user with email %q", a.User))
		}
		if err != nil {
			return err
		}
		target = u
	case a.Role != "":
		u, err := r.leastLoaded(ctx, job.OrgID, a.Role)
		if err != nil {
			return err
		}
		target = u
	default:
		return jobs.Permanent(errors.New("assign: action needs user or role"))
	}

	err = db.WithTx(ctx, r.d.Pool, func(tx pgx.Tx) error {
		if err := submissions.SetAssignee(ctx, tx, job.OrgID, ac.sub.ID, &target.ID); err != nil {
			return err
		}
		_, err := submissions.InsertEvent(ctx, tx, submissions.Event{
			SubmissionID: ac.sub.ID,
			OrgID:        job.OrgID,
			Type:         "assigned",
			ActorType:    "system",
			ActorName:    SystemActorName,
			Payload:      map[string]any{"assigneeId": target.ID.String(), "assigneeName": target.Name, "trigger": ac.payload.Trigger},
		})
		return err
	})
	if err != nil {
		return err
	}
	return r.recordSuccess(ctx, ac)
}

// leastLoaded picks, among users with role (or admins if none), the one with
// the fewest assigned non-terminal submissions; ties go to the earliest created.
func (r *runner) leastLoaded(ctx context.Context, orgID uuid.UUID, role string) (auth.User, error) {
	cands, err := r.d.Auth.UsersWithRole(ctx, orgID, role)
	if err != nil {
		return auth.User{}, err
	}
	if len(cands) == 0 {
		if cands, err = r.d.Auth.UsersWithRole(ctx, orgID, auth.RoleAdmin); err != nil {
			return auth.User{}, err
		}
	}
	if len(cands) == 0 {
		return auth.User{}, jobs.Permanent(fmt.Errorf("assign: no users with role %q or admin", role))
	}
	ids := make([]uuid.UUID, len(cands))
	for i, c := range cands {
		ids[i] = c.ID
	}
	rows, err := r.d.Pool.Query(ctx,
		`SELECT assignee_id, workflow_version_id, state FROM submissions
		 WHERE org_id = $1 AND assignee_id = ANY($2)`, orgID, ids)
	if err != nil {
		return auth.User{}, err
	}
	defer rows.Close()
	load := map[uuid.UUID]int{}
	for rows.Next() {
		var assignee uuid.UUID
		var wv *uuid.UUID
		var state string
		if err := rows.Scan(&assignee, &wv, &state); err != nil {
			return auth.User{}, err
		}
		terminal, err := r.isTerminal(ctx, wv, state)
		if err != nil {
			return auth.User{}, err
		}
		if !terminal {
			load[assignee]++
		}
	}
	if err := rows.Err(); err != nil {
		return auth.User{}, err
	}
	sort.SliceStable(cands, func(i, j int) bool {
		li, lj := load[cands[i].ID], load[cands[j].ID]
		if li != lj {
			return li < lj
		}
		return cands[i].CreatedAt.Before(cands[j].CreatedAt)
	})
	return cands[0], nil
}

func (r *runner) isTerminal(ctx context.Context, workflowVersionID *uuid.UUID, state string) (bool, error) {
	if workflowVersionID == nil {
		return true, nil // forms without a workflow end in "submitted", which is terminal
	}
	rec, err := r.d.Defs.GetWorkflowVersion(ctx, *workflowVersionID)
	if err != nil {
		return false, err
	}
	st, ok := rec.Definition.State(state)
	return ok && st.Terminal, nil
}

File: src/openforms/server/actions/assign.py
"""The assign action: a named user, or the least-loaded user with a role."""

from __future__ import annotations

import uuid
from typing import TYPE_CHECKING

import sqlalchemy as sa

from ..exceptions import NotFound
from ..models import auth as auth_models
from ..models import definitions as defs
from ..models import submissions as subs
from ..schemas.core import ROLE_ADMIN, User
from ..services.jobs import Job, Permanent
from .payload import SYSTEM_ACTOR_NAME

if TYPE_CHECKING:
    from sqlalchemy.ext.asyncio import AsyncSession

    from .runner import Runner


async def run_assign(r: Runner, job: Job) -> None:
    ac = await r.load(job)
    a = ac.payload.action
    async with r.d.db.session() as s:
        if a.user:
            try:
                target = await auth_models.get_user_by_email(s, job.org_id, a.user)
            except NotFound:
                raise Permanent(ValueError(f'assign: no user with email "{a.user}"')) from None
        elif a.role:
            target = await least_loaded(s, job.org_id, a.role)
        else:
            raise Permanent(ValueError("assign: action needs user or role"))
    async with r.d.db.transaction() as s:
        await subs.set_assignee(s, job.org_id, ac.sub.id, target.id)
        await subs.insert_event(
            s,
            subs.Event(
                submission_id=ac.sub.id,
                org_id=job.org_id,
                type=subs.EVENT_ASSIGNED,
                actor_type="system",
                actor_name=SYSTEM_ACTOR_NAME,
                payload={"assigneeId": str(target.id), "assigneeName": target.name, "trigger": ac.payload.trigger},
            ),
        )
    await r.record_success(ac)


async def least_loaded(s: AsyncSession, org_id: uuid.UUID, role: str) -> User:
    """Among users with ``role`` (or admins if none), the one with the fewest assigned
    non-terminal submissions; ties go to the earliest created."""
    cands = await auth_models.users_with_role(s, org_id, role) or await auth_models.users_with_role(
        s, org_id, ROLE_ADMIN
    )
    if not cands:
        raise Permanent(ValueError(f'assign: no users with role "{role}" or admin'))
    rows = await s.execute(
        sa.text("""SELECT assignee_id, workflow_version_id, state FROM submissions
        WHERE org_id = :org AND assignee_id = ANY(:ids)"""),
        {"org": org_id, "ids": [c.id for c in cands]},
    )
    load: dict[uuid.UUID, int] = {}
    for assignee, wv, state in rows:
        if not await _is_terminal(s, wv, state):
            load[assignee] = load.get(assignee, 0) + 1
    order = {c.id: i for i, c in enumerate(cands)}  # already ordered by created_at, id
    return min(cands, key=lambda c: (load.get(c.id, 0), c.created_at, order[c.id]))


async def _is_terminal(s: AsyncSession, workflow_version_id: uuid.UUID | None, state: str) -> bool:
    if workflow_version_id is None:
        return True  # forms without a workflow end in "submitted", which is terminal
    st = (await defs.get_workflow_version(s, workflow_version_id)).definition.state(state)
    return st is not None and st.terminal
